### DISCUSSION TITLE
Memory leak caused by `@validate_call` decorator

### DISCUSSION THREAD

**Discussion by ff137** [pydantic/pydantic#7912] [Pydantic V2]
Edit: Please see this follow-up post where cause of memory leak was isolated: https://github.com/pydantic/pydantic/discussions/7912#discussioncomment-7650465

Original post:
___

Good day,

We recently upgraded to using Pydantic v2 in our projects: [aries-cloudapi-python](https://github.com/didx-xyz/aries-cloudapi-python) and [aries-cloudcontroller-python](https://github.com/didx-xyz/aries-cloudcontroller-python).

DataDog provides us with a variety of metrics, and with it we notice that memory usage for Pydantic is remarkably high.

I've attached a sample flame graph of heap memory use for one of our methods, only showing the Pydantic trace. 
The picture is at the end of the post, as it's quite large.

The exact method being tested, `create_tenant`, can be [viewed here](https://github.com/didx-xyz/aries-cloudapi-python/blob/18b1bf82d50ad3195f2fa5e417f47b11612f4c43/app/routes/admin/tenants.py#L44).

To summarise how Pydantic is being used in this method being visualised:
- a [CreateTenantRequest](https://github.com/didx-xyz/aries-cloudapi-python/blob/18b1bf82d50ad3195f2fa5e417f47b11612f4c43/app/models/tenants.py#L19) model is used for our request body. A fairly straightforward BaseModel implementation.
- we then create a different Pydantic model: a `CreateWalletRequestWithGroups`.
  - This extends a `CreateWalletRequest`, which comes from our cloudcontroller library.
  - In short, this `CreateWalletRequest` is a BaseModel with custom validators, and it was autogenerated with the help of openapi-generator. Its definition can be [viewed here](https://github.com/didx-xyz/aries-cloudcontroller-python/blob/eddc27d7f631645bf0612aafb889a91c7f92fb65/aries_cloudcontroller/models/create_wallet_request.py#L31).

So, those are the two models being used in the method being shown.

If we look at the flame graph, we can see `simplify_schema_references` / `walk_core_schema` making a lot of recursive calls, involving count_refs, walk, handle_call_schema, handle_other_schemas, etc. The detail of what's going on here is very obscure and cryptic to me, but not too important.

Most important: the abovementioned `simplify_schema_references` / `walk_core_schema` pattern, is being called twice in our route method, and twice in `create_wallet (multitenancy_api.py)` (the right hand side of the flame graph). This obviously seems to represent some unnecessary overhead.

For clarity, this `create_wallet` method is being called [here](https://github.com/didx-xyz/aries-cloudapi-python/blob/18b1bf82d50ad3195f2fa5e417f47b11612f4c43/app/routes/admin/tenants.py#L75), with its implementation being in cloudcontroller [here](https://github.com/didx-xyz/aries-cloudcontroller-python/blob/eddc27d7f631645bf0612aafb889a91c7f92fb65/aries_cloudcontroller/api/multitenancy_api.py#L51).

The way I see it, we are creating two models in our `create_tenant` route, so it makes sense for the validation process to be called twice there, on the left hand side. The anomaly for me is that `simplify_schema_references` is being called 4 times in total ... and each time appears to be a rather expensive call, even with our more simple `CreateTenantRequest` model.

Note, `generate_schema` appears in the middle (purple column), and I'm guessing (I may be wrong) that is for the 2nd model we are creating, the one that was autogenerated and has additional validators / more complex definition. This purple column is very thin on the right hand side, so I assume this step not needed to run twice for one model.

FYI: the model config used for the `CreateWalletRequest`, autogenerated model is as follows:
```py
DEFAULT_PYDANTIC_MODEL_CONFIG: ConfigDict = {
    "defer_build": True,
    "populate_by_name": True,
    "validate_assignment": False,
}
```
We thought that defer_build: True and validate_assignment: False would help, but performance remained mostly the same. 

Now, the inefficient memory use doesn't make sense to me, and I'm creating this post because we please need help to understand how Pydantic can be used in a more effective and efficient way.

I understand this is a lengthy problem report, it's just for proper context. I don't expect any quick answers, but if you do have any insight to offer, or ideas for what we can try next, where we can test and provide further results, please let us know!

Thanks for your time!

DataDog flame graph of heap memory use by Pydantic, for our `create_tenant` method:

![Screenshot from 2023-10-24 17-22-25](https://github.com/pydantic/pydantic/assets/31511766/2013531a-c6a3-4000-ad48-6ceb053bb87b)

**Comment by rblaine95** [comment]
I'm a colleague of @ff137 and am working alongside him on this.

Please don't hesitate to ask for any additional information, I'll try my best to respond quickly ❤️

**Reply by rblaine95** [reply]
Here are some high resolution screenshots of the Datadog Live Heap Size Flame Graph.
The timespan is 25 October 2023, 12:15pm to 1:35pm.
**This is after a load test was run and the application was left idle for 1.5-2 hours.**

By Library (right hand side):
<img width="1512" alt="Screenshot 2023-10-25 at 1 50 13 PM" src="https://github.com/pydantic/pydantic/assets/4052340/3fdcb6f5-6da3-4aaf-9238-1588eff30127">

By Function (right hand side):
<img width="1512" alt="Screenshot 2023-10-25 at 1 51 15 PM" src="https://github.com/pydantic/pydantic/assets/4052340/dc46f5b1-589b-4c22-b007-6ed4e59b5724">

**Comment by ff137** [comment]
I've been able to isolate the cause of the memory leak... it comes down to the `@validate_call` decorator.

@samuelcolvin I'm interested to hear if this is a known issue

Here's a basic script to reproduce the memory leak issue:
```py
import asyncio
import tracemalloc
from typing import Any, Dict, List, Optional, Tuple, Union
from uuid import uuid4

from aries_cloudcontroller import (
    AcaPyClient,
    CreateWalletRequest,
    CreateWalletResponse,
    MultitenancyApi,
)
from pydantic import BaseModel, Field, StrictFloat, StrictInt, StrictStr, validate_call
from typing_extensions import Annotated


# Our custom request model
class CreateTenantRequest(BaseModel):
    wallet_label: str
    wallet_name: Optional[str] = None
    roles: Optional[List[str]] = None
    group_id: Optional[str] = None
    image_url: Optional[str] = None


# stub the MultitenancyApi module that is called in cloudcontroller, with identical method signature
class StubbedMultitenancyApi(MultitenancyApi):
    @validate_call  # memory leak disappears if this is commented out
    async def create_wallet(
        self,
        body: Optional[CreateWalletRequest] = None,
        _request_timeout: Union[
            None,
            Annotated[StrictFloat, Field(gt=0)],
            Tuple[
                Annotated[StrictFloat, Field(gt=0)], Annotated[StrictFloat, Field(gt=0)]
            ],
        ] = None,
        _request_auth: Optional[Dict[StrictStr, Any]] = None,
        _content_type: Optional[StrictStr] = None,
        _headers: Optional[Dict[StrictStr, Any]] = None,
        _host_index: Annotated[StrictInt, Field(ge=0, le=0)] = 0,
    ) -> CreateWalletResponse:
        pass


# patch the AcaPyClient to use the stubbed MultitenancyApi
class PatchedAcaPyClient(AcaPyClient):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # Replace the standard MultitenancyApi with our patched version
        self.multitenancy = StubbedMultitenancyApi(self.api_client)


# Use patched client to call `create_wallet`
async def simulate_cloudcontroller_api_call(data):
    wallet_request = CreateWalletRequest(**data)  # convert our custom request model to cloudcontroller model
    async with PatchedAcaPyClient("nohostneeded", admin_insecure=True) as client:
        await client.multitenancy.create_wallet(wallet_request)


async def create_tenant(data):
    tenant_request = CreateTenantRequest(**data)  # populate our custom request model
    wallet_data = {
        "image_url": tenant_request.image_url,
        "key_management_mode": "managed",
        "label": tenant_request.wallet_label,
        "wallet_key": "some_random_key",
        "wallet_name": tenant_request.wallet_name or uuid4().hex,
        "wallet_type": "askar",
    }
    await simulate_cloudcontroller_api_call(wallet_data)


# call `create_tenant` n times and print memory stats
async def main():
    tracemalloc.start()

    n = 1000  # create n tenants

    # Dummy request data
    data = {
        "wallet_label": "wallet1",
        "roles": ["issuer"],
        "image_url": "http://example.com/image.png",
    }

    # Loop to simulate repeated calls
    for _ in range(n):
        await create_tenant(data)

    print(f"Memory usage after {n} tenants created:")
    snapshot = tracemalloc.take_snapshot()
    top_stats = snapshot.statistics("lineno")
    for stat in top_stats[:5]:
        print(stat)


asyncio.get_event_loop().run_until_complete(main())
```

After installing the cloudcontroller dependency (`pip install aries-cloudcontroller`), you can run the script (`python script.py`).
(Note: also install latest pydantic)

The above script stubs the original cloudcontroller API calls that would be made with the `create_tenant` request.

Memory stats are printed at the end of the script's runtime as follows:
```py
Memory usage after 1000 tenants created:
../lib/python3.9/site-packages/pydantic/_internal/_core_utils.py:210: size=11.5 MiB, count=71673, average=168 B
../lib/python3.9/site-packages/pydantic/_internal/_core_utils.py:585: size=7473 KiB, count=105985, average=72 B
../lib/python3.9/site-packages/pydantic/_internal/_core_utils.py:353: size=1801 KiB, count=10988, average=168 B
../lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:689: size=1001 KiB, count=16014, average=64 B
../lib/python3.9/site-packages/aiohttp/client.py:317: size=984 KiB, count=1000, average=1008 B
```

As you can see, pydantic is holding on to memory. Each of those _core_utils.py code lines points to somewhere that `.copy()` or validation is called.

If you run the above script again, but this time commenting out the `@validate_call` decorator, you get:
```py
Memory usage after 1000 tenants created:
../lib/python3.9/site-packages/pydantic/_internal/_core_utils.py:210: size=17.2 KiB, count=114, average=154 B
../lib/python3.9/site-packages/pydantic/_internal/_core_utils.py:585: size=4496 B, count=64, average=70 B
../lib/python3.9/site-packages/pydantic/_internal/_core_utils.py:353: size=4056 B, count=26, average=156 B
../lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:2009: size=1936 B, count=19, average=102 B
../lib/python3.9/site-packages/pydantic/_internal/_generate_schema.py:689: size=1848 B, count=24, average=77 B
```

Then there is no excessive holding on to memory!

Ultimately, it is coming down to model validation being called for each request (on a complex model), and for some reason this is not garbage collected. Adding manual `gc.collect()` calls in the override methods, or things like `del body`, does not solve it.

Fortunately, I think we can get away without using the `@validate_call` method, because our API already does model validation upon receiving requests, before forwarding it to the AcaPyClient, so the secondary validation is probably redundant.

But this of course unexpected behaviour. @samuelcolvin is this a known issue, with any possible workarounds? Or have we stumbled on something new?
